docs/console-access.md

```diff
@@ -0,0 +1,340 @@
+Console access for virtual machines
+===================================
+
+There are two main ways tenant users may interact with virtual machines they
+are running.
+
+ * Guest OS hosted network services. This covers SSH server, web apps like
+   cockpit, Windows remote desktop. The set of services available varies
+   according to what OS is installed, and can only be accessed if the guest
+   is running "normally".
+
+ * Host OS hosted network services. This covers VNC, SPICE, and serial port
+   console services exposed over the host network by the QEMU emulator. These
+   are available as long as the virtual machine is powered on, regardless of
+   whether the guest OS is running "normally" or not. The set of services
+   available is the same regardless of what kind of guest OS is installed.
+
+Access to guest OS hosted network services does not require any special work
+by kubevirt aside from ensuring the guest has a working network connection to
+the tenant. This document thus considers how to expose the host OS hosted
+network services to tenant users.
+
+
+Proxied vs direct access
+------------------------
+
+The VNC, SPICE or serial console network service exposed by QEMU can be made
+to listen on either a local-only UNIX domain socket, or a remotely accessible
+TCP socket. These choices in QEMU configuration, lead to a number of different
+options for how a tenant user can be allowed to connect.
+
+If the QEMU network service listens on a UNIX domain socket, then there will
+need to be a proxy service running on each compute node that exposes the
+service over TCP for remote access. There are several options for proxying, but
+the common approach is to use either a traditional HTTP CONNECT based proxy, or
+a HTTP websockets encoded proxy. In such a scenario, the compute host will only
+need to expose a single TCP port, no matter how many virtual machines are
+running. If tenant users cannot access IP addresses in the cluster network, then
+the POD running the proxy on each host will need an associated service IP
+address, to which tenant users will connect.
+
+If the QEMU network service listens on a TCP socket, then each virtual machine
+running on a compute host will be listening on the same IP address, but using a
```

Comment:
I meant to say, where I think virt-api coming into play is providing a way to query the address (host+port) of the console(s) (there can be multiple text consoles per vm), and acquire security token(s) to access them. The actual establishment of the connection to the console would be done by the client app directly.